File: archive/programmers/42583/ppoddo/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    bridge = deque([0] * bridge_length)
    trucks = deque(truck_weights)
    current_bridge_weight = 0
    current_time = 0

    while trucks or current_bridge_weight > 0:
        current_time += 1
        out = bridge.popleft()
        current_bridge_weight -= out
        if trucks and current_bridge_weight + trucks[0] <= weight:
            t = trucks.popleft()
            bridge.append(t)
            current_bridge_weight += t
        else:
            bridge.append(0)

    return current_time

File: archive/programmers/42583/haeun/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    let time = bridge_length;
    const stack = [];
    let currWeight = 0;
    
    for(const w of truck_weights) {
       if(!stack.length) {
           stack.push(w);
           currWeight += w;
           time++;
           continue;
       }
        
        while (1) {
            time++;
            
            if(stack.length === bridge_length) 
                currWeight -= stack.shift();
            
            const isWeightOver = currWeight + w > weight;
            
            if(isWeightOver) stack.push(0);
            else {
                stack.push(w);
                currWeight += w;
                break;
            }
        }
    }
    
    return time;
}